Bio/bgzf.py

```diff
@@ -852,33 +871,58 @@ def _write_block(self, block):
+        if len(data_bytes) == len(data_str):
+            # fast case -- 8-bit only
+            self._write_bytes(data_bytes)
+            return
+
+        # When reading we can't cope with multi-byte UTF-8 characters
+        # being split between BGZF blocks. Therefore, if the 64KB block
+        # end "in the middle" of a character, the block is resized down
+        # to the end of the previous character. This can produce block
+        # up to 3 bytes smaller than the full 64KB.
+        assert self._encoding == "utf-8"
+        assert len(self._buffer) < self._BLOCK_SIZE
+        while len(data_bytes):
+            safe_size = self._BLOCK_SIZE - len(self._buffer)
+            if safe_size >= len(data_bytes):
+                safe_size = len(data_bytes)
+            else:
+                while safe_size > 0 and data_bytes[safe_size] >> 6 == 2:
+                    # the byte after the block boundary is not the first octet
+                    # of UTF-8 character (RFC 3629)
+                    safe_size -= 1
+            if len(self._buffer):
+                self._buffer += data_bytes[:safe_size]
+            else:
+                self._buffer = data_bytes[:safe_size]  # avoid copy
+            data_bytes = data_bytes[safe_size:]
+            if len(data_bytes) == 0 and len(self._buffer) < self._BLOCK_SIZE:
+                # do not write the last non-full block out
+                break
+            self._write_block(self._buffer)
+            self._buffer = b""
+
     def write(self, data):
         """Write method for the class."""
-        # TODO - Check bytes vs unicode
         if isinstance(data, str):
-            # When reading we can't cope with multi-byte characters
-            # being split between BGZF blocks, so we restrict to a
-            # single byte encoding - like ASCII or latin-1.
-            # On output we could probably allow any encoding, as we
-            # don't care about splitting unicode characters between blocks
-            data = data.encode("latin-1")
-        # block_size = 2**16 = 65536
-        data_len = len(data)
-        if len(self._buffer) + data_len < 65536:
-            # print("Cached %r" % data)
-            self._buffer += data
+            self._write_str(data)
         else:
-            # print("Got %r, writing out some data..." % data)
-            self._buffer += data
-            while len(self._buffer) >= 65536:
-                self._write_block(self._buffer[:65536])
-                self._buffer = self._buffer[65536:]
+            self._write_bytes(data)
 
     def flush(self):
-        """Flush data explicitally."""
-        while len(self._buffer) >= 65536:
-            self._write_block(self._buffer[:65535])
-            self._buffer = self._buffer[65535:]
+        """Flush data explicitly."""
+        assert len(self._buffer) < self._BLOCK_SIZE
+        # TODO is writing an empty buffer/block an intended behavior?
```

Comment:
From memory yes, that is intended.

And again, explicit exceptions preferred over asserts.